File: apis_c#/AthleteApi/AthleteApi/Services/TournamentService.cs
using System.Data;
using Microsoft.Data.SqlClient;
using AthleteApi.Models;

namespace AthleteApi.Services
{
    // Implementación del servicio de torneos que maneja las operaciones CRUD para los torneos
    public class TournamentService : ITournamentService
    {
        private readonly string _connectionString;

        // Constructor que inicializa la cadena de conexión a la base de datos
        public TournamentService(IConfiguration configuration)
        {
            // Obtiene la cadena de conexión desde la configuración
            _connectionString = configuration?.GetConnectionString("DefaultConnection") ?? throw new ArgumentNullException(nameof(configuration));
        }

        // Método para crear un nuevo torneo en la base de datos
        public async Task CreateTournament(Tournament tournament)
        {
            // Crea una nueva conexión SQL usando la cadena de conexión
            using (SqlConnection conn = new SqlConnection(_connectionString))
            {
                // Crea un nuevo comando SQL para ejecutar el procedimiento almacenado 'sp_createTournament'
                using (SqlCommand cmd = new SqlCommand("sp_createTournament", conn))
                {
                    cmd.CommandType = CommandType.StoredProcedure;

                    // Agrega los parámetros necesarios para el procedimiento almacenado
                    cmd.Parameters.AddWithValue("@Name", tournament.Name);
                    cmd.Parameters.AddWithValue("@Location", tournament.Location);
                    cmd.Parameters.AddWithValue("@StartDate", tournament.StartDate);
                    cmd.Parameters.AddWithValue("@EndDate", tournament.EndDate);

                    // Abre la conexión a la base de datos
                    conn.Open();
                    // Ejecuta el comando de forma asíncrona
                    await cmd.ExecuteNonQueryAsync();
                }
            }
        }

        // Método para obtener una lista de torneos desde la base de datos con paginación y filtro opcional por nombre
        public async Task<IEnumerable<Tournament>> GetTournaments(
            int pageNumber, int pageSize, string? name = null)
        {
            var tournaments = new List<Tournament>();

            // Crea una nueva conexión SQL usando la cadena de conexión
            using (SqlConnection conn = new SqlConnection(_connectionString))
            {
                // Crea un nuevo comando SQL para ejecutar el procedimiento almacenado 'sp_getTournaments'
                using (SqlCommand cmd = new SqlCommand("sp_getTournaments", conn))
                {
                    cmd.CommandType = CommandType.StoredProcedure;

                    // Agrega los parámetros necesarios para el procedimiento almacenado
                    cmd.Parameters.AddWithValue("@PageNumber", pageNumber);
                    cmd.Parameters.AddWithValue("@PageSize", pageSize);
                    cmd.Parameters.AddWithValue("@Name", name ?? (object)DBNull.Value);

                    // Abre la conexión a la base de datos
                    conn.Open();
                    // Ejecuta el comando y obtiene un lector de datos de forma asíncrona
                    using (SqlDataReader reader = await cmd.ExecuteReaderAsync())
                    {
                        // Lee los datos devueltos por el lector de datos
                        while (await reader.ReadAsync())
                        {
                            var tournament = new Tournament
                            {
                                Id = reader.GetInt32(0),
                                Name = reader.GetString(1),
                                Location = reader.GetString(2),
                                StartDate = reader.GetDateTime(3),
                                EndDate = reader.GetDateTime(4)
                            };
                            // Agrega el torneo a la lista de torneos
                            tournaments.Add(tournament);
                        }
                    }
                }
            }

            // Retorna la lista de torneos
            return tournaments;
        }
    }
}

File: apis_py/app/services/tournament_service.py
from typing import List, Optional
from fastapi import Depends
from sqlalchemy.orm import Session
from app.models.tournament import Tournament
from app.schemas.tournament import TournamentCreate, TournamentResponse
from app.database import get_db

class TournamentService:
    """
    Servicio para manejar operaciones relacionadas con torneos en la base de datos.
    
    Atributos:
        db (Session): Sesión de la base de datos para realizar las operaciones.
    """

    def __init__(self, db: Session = Depends(get_db)):
        """
        Inicializa el servicio con una sesión de la base de datos.
        
        Args:
            db (Session): Sesión de la base de datos inyectada automáticamente por FastAPI.
        """
        self.db = db

    def get_tournaments(self, page: int, size: int, id: Optional[int], name: Optional[str]) -> List[TournamentResponse]:
        """
        Obtiene una lista paginada de torneos desde la base de datos, filtrando por ID y nombre si se proporcionan.

        Args:
            page (int): Número de la página a obtener.
            size (int): Cantidad de registros por página.
            id (Optional[int]): Filtro por ID del torneo.
            name (Optional[str]): Filtro por nombre del torneo.

        Returns:
            List[TournamentResponse]: Lista de respuestas con la información de los torneos.
        """
        query = self.db.query(Tournament)
        if id:
            query = query.filter(Tournament.id == id)
        if name:
            query = query.filter(Tournament.name == name)
        query = query.order_by(Tournament.id)
        tournaments = query.offset((page - 1) * size).limit(size).all()
        return [TournamentResponse.from_orm(tournament) for tournament in tournaments]

    def create_tournament(self, tournament: TournamentCreate) -> TournamentResponse:
        """
        Crea un nuevo torneo en la base de datos.

        Args:
            tournament (TournamentCreate): Datos del torneo a crear.

        Returns:
            TournamentResponse: Respuesta con la información del torneo creado.
        """
        db_tournament = Tournament(**tournament.dict())
        self.db.add(db_tournament)
        self.db.commit()
        self.db.refresh(db_tournament)
        return TournamentResponse.from_orm(db_tournament)
